source/server/options_impl.cc

```diff
@@ -125,18 +137,20 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
   } else if (mode.getValue() == "validate") {
     mode_ = Server::Mode::Validate;
   } else {
-    std::cerr << "error: unknown mode '" << mode.getValue() << "'" << std::endl;
-    exit(1);
+    const std::string message = fmt::format("error: unknown mode '{}'", mode.getValue());
+    std::cerr << message << std::endl;
+    throw MalformedArgvException(message);
   }
 
   if (local_address_ip_version.getValue() == "v4") {
     local_address_ip_version_ = Network::Address::IpVersion::v4;
   } else if (local_address_ip_version.getValue() == "v6") {
     local_address_ip_version_ = Network::Address::IpVersion::v6;
   } else {
-    std::cerr << "error: unknown IP address version '" << local_address_ip_version.getValue() << "'"
-              << std::endl;
-    exit(1);
+    const std::string message =
+        fmt::format("error: unknown IP address version '{}'", local_address_ip_version.getValue());
+    std::cerr << message << std::endl;
```

Comment:
One minor consideration might be that we could consider pushing these `std::cerr` statements out to where the exception is caught. I.e. make `OptionalImpl` pure and the surrounding code responsible for the icky world of stderr I/O and exiting.

Reply:
That's actually where I started, but it's awkward because of the cases where tclap does the printing. For example if you pass a bad flag name (`envoy --asdf`), in addition to the error message tclap helpfully prints a usage string. We can stop it doing that, but I think we want to keep it -- and unless we want to expose tclap beyond this class, that means doing it here.

So our choices are either to do the stderr work always or sometimes, and reluctantly I think "always" is better.